presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotSegmentPageSourceForGrpc.java

```diff
@@ -0,0 +1,142 @@
+            serverResponseIterator = queryPinot(split);
+        }
+        ByteBuffer byteBuffer = null;
+        if (serverResponseIterator.hasNext()) {
+            long startTimeNanos = System.nanoTime();
+            ServerResponseWrapper serverResponse = serverResponseIterator.next();
+            readTimeNanos += System.nanoTime() - startTimeNanos;
+            switch (serverResponse.getResponseType()) {
+                case Constants.Response.ResponseType.DATA:
+                    estimatedMemoryUsageInBytes = serverResponse.getSerializedSize();
+                    // Store each dataTable which will later be constructed into Pages.
+                    try {
+                        byteBuffer = serverResponse.getPayloadReadOnlyByteBuffer();
+                        DataTable dataTable = serverResponse.getDataTable(byteBuffer);
+                        checkExceptions(dataTable, split, PinotSessionProperties.isMarkDataFetchExceptionsAsRetriable(session));
+                        currentDataTable = new PinotSegmentPageSource.PinotDataTableWithSize(dataTable, serverResponse.getSerializedSize());
+                    }
+                    catch (IOException e) {
+                        throw new PinotException(
+                            PINOT_DATA_FETCH_EXCEPTION,
+                            split.getSegmentPql(),
+                            String.format("Encountered Pinot exceptions when fetching data table from %s: %s", split, e));
+                    }
+                    break;
+                case Constants.Response.ResponseType.METADATA:
+                    // The last part of the response is Metadata
+                    currentDataTable = null;
+                    serverResponseIterator = null;
+                    close();
+                    return null;
+                default:
+                    throw new PinotException(
+                        PINOT_UNEXPECTED_RESPONSE,
+                        split.getSegmentPql(),
+                        String.format("Encountered Pinot exceptions, unknown response type - %s", serverResponse.getResponseType()));
+            }
+        }
+        Page page = fillPage();
+        completedPositions += currentDataTable.getDataTable().getNumberOfRows();
+        if (byteBuffer != null) {
+            byteBuffer.clear();
+        }
+        return page;
+    }
+
+    private Iterator<ServerResponseWrapper> queryPinot(PinotSplit split)
+    {
+        String sql = split.getSegmentPql().orElseThrow(() -> new PinotException(PINOT_INVALID_PQL_GENERATED, Optional.empty(), "Expected the segment split to contain the pql"));
+        String host = split.getSegmentHost().orElseThrow(() -> new PinotException(PINOT_INVALID_PQL_GENERATED, Optional.empty(), "Expected the segment split to contain the host"));
+        String[] hostSplits = host.split("_");
+        String grpcHost = (hostSplits.length > 1) ? hostSplits[hostSplits.length - 2] : host;
```

Comment:
I didn't follow this part: Does the host in the split now contain two hostnames ? The GRPC one and the regular one ? An example of the response returned by the server would be helpful in the code comments.

Can this stuff be done by the split manager when creating the splits ? Ideally you want the splits to be very definitive about what behavior will happen at execution time.

Reply:
pinot server instance is named as `Server_${hostname}_${port}`, however this port is not gRpc port.
Hence we need to use this instance name to query Pinot controller for the instance config which contains the gRpc port.

I can embed this logic to split.